src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs

```diff
@@ -0,0 +1,465 @@
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="JsonObject"/> class representing the empty object.
+        /// </summary>
+        /// <param name="duplicatePropertyNameHandling">Specifies the way of handling duplicate property names.</param>
+        /// <exception cref="ArgumentException">
+        ///   Provided manner of handling duplicates does not exist.
+        /// </exception>
+        public JsonObject(DuplicatePropertyNameHandling duplicatePropertyNameHandling = default)
+        {
+            if(!Enum.IsDefined(typeof(DuplicatePropertyNameHandling), duplicatePropertyNameHandling))
+            {
+                throw new ArgumentException(SR.InvalidDuplicatePropertyNameHandling);
+            }
+
+            _dictionary = new Dictionary<string, JsonNode>();
+            _duplicatePropertyNameHandling = duplicatePropertyNameHandling;
+        }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="JsonObject"/> class representing provided set of JSON properties.
+        /// </summary>
+        /// <param name="jsonProperties">>Properties to represent as a JSON object.</param>
+        /// <param name="duplicatePropertyNameHandling">Specifies the way of handling duplicate property names.</param>
+        /// <exception cref="ArgumentException">
+        ///   Provided collection contains duplicates if handling duplicates is set to <see cref="DuplicatePropertyNameHandling.Error"/>.
+        /// </exception>
+        public JsonObject(IEnumerable<KeyValuePair<string, JsonNode>> jsonProperties, DuplicatePropertyNameHandling duplicatePropertyNameHandling = default)
+            : this(duplicatePropertyNameHandling)
+            => AddRange(jsonProperties);
+
+        /// <summary>
+        ///   Gets or sets the value of the specified property.
+        /// </summary>
+        /// <param name="propertyName">The property name of the value to get or set.</param>
+        /// <exception cref="ArgumentNullException">
+        ///   Provided property name is null.
+        /// </exception>
+        public JsonNode this[string propertyName]
+        {
+            get => propertyName != null ? GetPropertyValue(propertyName) : throw new ArgumentNullException(nameof(propertyName));
+            set => _dictionary[propertyName] = value ?? throw new ArgumentNullException(nameof(propertyName));
+        }
+
+        /// <summary>
+        ///   Returns an enumerator that iterates through the JSON object properties.
+        /// </summary>
+        /// <returns>An enumerator structure for the JSON object.</returns>
+        /// <exception cref="ArgumentException">
+        ///   Property name to set already exists if handling duplicates is set to <see cref="DuplicatePropertyNameHandling.Error"/>.
+        /// </exception>
+        public IEnumerator<KeyValuePair<string, JsonNode>> GetEnumerator() => _dictionary.GetEnumerator();
```

Comment:
The generic dictionary uses a strongly-typed (struct) return here for performance reasons.

If we think that there's going to be a lot of reading happening from these types (particularly, this enumerator) then this type probably wants to return its own strongly typed enumerator (one which just wraps the one from the dictionary).